src/ecs.rs

```diff
@@ -0,0 +1,692 @@
+            // TODO: make best guesses? right now these should rarely happen at info and debug level and user can switch to forest if needed
+        }
+        Ok(())
+    }
+
+    /// Parses a [`Span`] into an [`EcsEvent`]. Barely implemented.
+    fn parse_span(
+        span: &Span,
+        duration_root: Option<f64>,
+        ecs_event: &mut EcsEvent,
+    ) -> fmt::Result {
+        let total_duration = span.total_duration().as_nanos() as f64;
+        let root_duration = duration_root.unwrap_or(total_duration);
+
+        for tree in span.nodes() {
+            EcsFormatter::parse_tree(tree, Some(root_duration), ecs_event)?;
+        }
+
+        Ok(())
+    }
+}
+
+/// Quote spec for parsing the quoted strings emitted as field values by tracing.
+#[derive(Copy, Clone, Debug)]
+struct TraceQuoteSpec;
+
+impl quoted_string::spec::GeneralQSSpec for TraceQuoteSpec {
+    type Quoting = Self;
+    type Parsing = TraceParsingImpl;
+}
+
+impl quoted_string::spec::QuotingClassifier for TraceQuoteSpec {
+    fn classify_for_quoting(pcp: PartialCodePoint) -> QuotingClass {
+        if !is_valid_pcp(pcp) {
+            QuotingClass::Invalid
+        } else {
+            match pcp.as_u8() {
+                b'"' | b'\\' => QuotingClass::NeedsQuoting,
+                _ => QuotingClass::QText,
+            }
+        }
+    }
+}
+
+fn is_valid_pcp(pcp: PartialCodePoint) -> bool {
+    let bch = pcp.as_u8();
+    (b' '..=b'~').contains(&bch)
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug, Hash)]
+struct TraceParsingImpl;
+
+impl quoted_string::spec::ParsingImpl for TraceParsingImpl {
+    fn can_be_quoted(pcp: PartialCodePoint) -> bool {
+        is_valid_pcp(pcp)
+    }
+    fn handle_normal_state(
+        pcp: PartialCodePoint,
+    ) -> Result<(quoted_string::spec::State<Self>, bool), quoted_string::error::CoreError> {
+        if is_valid_pcp(pcp) {
+            Ok((quoted_string::spec::State::Normal, true))
+        } else {
+            Err(quoted_string::error::CoreError::InvalidChar)
+        }
+    }
+
+    fn advance(
+        &self,
+        pcp: PartialCodePoint,
+    ) -> Result<(quoted_string::spec::State<Self>, bool), quoted_string::error::CoreError> {
+        if is_valid_pcp(pcp) {
+            Ok((quoted_string::spec::State::Normal, false))
+        } else {
+            Err(quoted_string::error::CoreError::InvalidChar)
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Default, Debug)]
+pub(crate) struct EcsEvent {
+    /// Date/time when the event originated, which will be formatted as a string in RFC3339 format.
+    #[serde(
+        default = "default_timestamp_now",
+        rename(serialize = "@timestamp", deserialize = "@timestamp")
+    )]
+    timestamp: DateTime<Utc>,
+    /// The primary message to be logged.
+    message: String,
+    /// Details about the event’s logging mechanism or logging transport.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    log: Option<EcsLogFieldSet>,
```

Comment:
Virtually everything in ECS is optional except for the timestamp. Other platforms also tend to expect a message field, so that's de facto required, but everything else is optional. Intuition suggests it would be very weird to include empty objects that don't contain anything to be indexed in a context where bytes processed are often how log processing is billed.